### PR TITLE
istioctl: add command to manage logging levels for istiod

### PR DESCRIPTION
**Pull request by piotrkpc**
Currently there is `istioctl pc log` command for Envoy proxy.
This changes adds similar functionalities for `istiod` through
`controlz` API.

Implements #24922

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ x ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

**Comment by piotrkpc** [comment]
Still need to implement the basic get functionalities and probably should refactor those if's in command func.

File: istioctl/cmd/istiodconfig.go [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+		Use:   "log [--level <scope>:<level>][--stack-trace-level <scope>:<level>]|[-r|--reset]",
+		Short: "Manage istiod logging.",
+		Long:  "Retrieve or update logging levels of istiod components.",
+		Example: ` # Retrieve information about istiod logging levels.
+  istioctl istiod-config log [<pod-name>]
+
+  # Update levels of the specified loggers.
+  istioctl istiod-config log --level ads:debug,authorization:debug
+
+  # Reset levels of all the loggers to default value (warning).
```
Comment (howardjohn) [inline istioctl/cmd/istiodconfig.go:152]: nit: warning is not default log level for istiod

File: istioctl/cmd/istiodconfig.go [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+func newScopeLevelPair(slp, validationPattern string) (*scopeLevelPair, error) {
+
+	matched, err := regexp.MatchString(validationPattern, slp)
+	if err != nil {
+		return nil, err
+	}
+	if !matched {
+		return nil, fmt.Errorf("pattern %s did not match", slp)
+	}
+	fmt.Println("validation ok ", slp)
```
Comment (howardjohn) [inline istioctl/cmd/istiodconfig.go:43]: Probably don't want this?

Comment (piotrkpc) [reply istioctl/cmd/istiodconfig.go:43]: sure

File: istioctl/cmd/istiodconfig.go (hunk 1 above)
Comment (piotrkpc) [reply istioctl/cmd/istiodconfig.go:152]: good catch thank you.

**Comment by piotrkpc** [comment]
I've refactored the code a bit and added more tests. I think it's ready for reviews.

File: istioctl/cmd/istiodconfig.go [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+			client, err := kubeClientWithRevision(kubeconfig, configContext, opts.Revision)
+			if err != nil {
+				return fmt.Errorf("failed to create k8s client: %v", err)
+			}
+
+			var podName, ns string
+			if len(args) == 0 {
+				pl, err := client.PodsForSelector(context.TODO(), handlers.HandleNamespace(istioNamespace, defaultNamespace), istiodLabelSelector)
+				if err != nil {
+					return fmt.Errorf("not able to locate pod with selector %s: %v", istiodLabelSelector, err)
```
Comment (howardjohn) [inline istioctl/cmd/istiodconfig.go:377]: help text says `istioctl istiod-config log [<pod-name>]`, but code says label selector. IMO, it should be either (or both):
* Take pod name, set log level on that one. Prefer this to match the proxy-config command, you may want to debug single instance
* Take label, set log level on all of them

IMO we should not use label selector then only use one of them

File: istioctl/cmd/istiodconfig.go [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+`,
+		Aliases: []string{"o"},
+		Args: func(logCmd *cobra.Command, args []string) error {
+			if istiodReset && outputLogLevel != "" {
+				logCmd.Println(logCmd.UsageString())
+				return fmt.Errorf("--level cannot be combined with --reset")
+			}
+			if istiodReset && stackTraceLevel != "" {
+				logCmd.Println(logCmd.UsageString())
+				return fmt.Errorf("--stack-trace-level cannot be combined with --reset")
```
Comment (howardjohn) [inline istioctl/cmd/istiodconfig.go:363]: Just throwing it out there, my 2c is that stack trace level is not very useful. Perfectly happy to include it, but I would be fine with removing it as well if we wanted to

File: istioctl/cmd/istiodconfig.go [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+	return "", nil
+}
+
+type logLevelState struct {
+	client            *ControlzClient
+	outputLogLevel    string
+	validationPattern string
+}
+
+func (ll *logLevelState) run() (string, error) {
```
Comment (howardjohn) [inline istioctl/cmd/istiodconfig.go:96]: Seems like this and reset and stackTraceLevelState should all be calling the same function and just passing in what to set. That way code is re-used - currently its not only repeated but also different; for some reason reset is parallel but this is not?

File: istioctl/cmd/istiodconfig.go (hunk 3 above)
Comment (piotrkpc) [reply istioctl/cmd/istiodconfig.go:377]: Square brackets around pod-name suppose to suggest it's optional parameter. Currently user can provide podname or selector or neither (default selector app=istiod will be used). Currently if there are multiple pods with label app=istiod only first one will used as in:
https://github.com/istio/istio/blob/d89c6a624cfd0962e2ed25238ec671baebda0ca5/istioctl/cmd/dashboard.go#L294-L296
I will improve the first line (Usage) of help for log subcommand

Comment (piotrkpc) [reply istioctl/cmd/istiodconfig.go:377]: Square brackets around pod-name suppose to suggest it's optional parameter. Currently user can provide podname or selector or neither (default selector app=istiod will be used). Currently if there are multiple pods with label app=istiod only first one will used as in:

istio/istioctl/cmd/dashboard.go

Lines 294 to 296 in d89c6a6
 if len(pl.Items) > 1 { 
 	log.Warnf("more than 1 pods fits selector: %s; will use pod: %s", labelSelector, pl.Items[0].Name) 
 } 

I will improve the first line (Usage) of help for log subcommand

File: istioctl/cmd/istiodconfig.go (hunk 4 above)
Comment (piotrkpc) [reply istioctl/cmd/istiodconfig.go:363]: I can remove it if this adds complexity

File: istioctl/cmd/istiodconfig.go (hunk 5 above)
Comment (piotrkpc) [reply istioctl/cmd/istiodconfig.go:96]: Thanks for feedback. I've refactored this a bit moving api calls concurrent to controlz Client

**Comment by piotrkpc** [comment]
@howardjohn I've introduced some improvements. I'll appreciate some reviews.

**Comment by howardjohn** [comment]
/ok-to-test

@istio/wg-user-experience-maintainers @esnible PTAL

**Review by esnible — APPROVED** [review]

**Comment by piotrkpc** [comment]
I'll update the release notes today.

> @piotrkpc: The following test **failed**, say `/retest` to rerun all failed tests:
> 
> Test name	Commit	Details	Rerun command
> release-notes_istio	[a3b34e9](https://github.com/istio/istio/commit/a3b34e984b132492fb4f2d71b47b77211e1c7e60)	[link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25276/release-notes_istio/3406)	`/test release-notes_istio`
> Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md). If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).

**Review by brian-avery — APPROVED** [review]

**Pull request merged**